web-common/src/features/entity-management/file-artifacts.ts

```diff
@@ -205,6 +212,13 @@ export class FileArtifacts {
    * Map of all files and its individual store
    */
   private readonly artifacts: Record<string, FileArtifact> = {};
+  /**
+   * A cache of {@Link FileArtifact}s by {@link ResourceKind}
+   * This avoids going through every file artifact to filter the by kind
+   */
+  private readonly artifactsByKind: Partial<
+    Record<ResourceKind, FileArtifact[]>
+  > = {};
```

Comment:
IMO it significantly increases the code complexity to have two sources-of-truth that need to be synchronized. E.g. needing to pass `handleArtifactNameChange` to the `FileArtifact` constructor adds an unfortunate coupling between the contained item and the container.

Can we remove the `artifactsByKind` cache and instead add a method like this:
```
 public getArtifactsByKind(kind: ResourceKind): FileArtifact[] {
    return Object.values(this.artifacts).filter(artifact => {
      const name = get(artifact.name);
      return name?.kind === kind;
    });
  }
```

The performance impact of dynamic filtering shouldn't be perceived by a user until the project has (many) thousands of files.

Reply:
Ya makes sense.